### PR TITLE
Code coverage for smoke tests.

### LINKED ISSUE #673: coverage reporting for smoke tests
It would be great to get coverage reports for the golang code exercised by our smoke tests.

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #673

### CODE REVIEW

File: test/140_weave_local_test.sh [hunk 1]
```diff
@@ -4,15 +4,15 @@
...
-run_on $HOST1 sudo weave --local launch --ipalloc-range 10.2.5.0/24
+run_on $HOST1 sudo COVERAGE=$COVERAGE weave --local launch --ipalloc-range 10.2.5.0/24
 assert_raises "docker_on $HOST1 ps | grep weave"
 
-run_on $HOST1 sudo weave --local run 10.2.6.5/24 -ti --name=c1 $SMALL_IMAGE /bin/sh
+run_on $HOST1 sudo COVERAGE=$COVERAGE weave --local run 10.2.6.5/24 -ti --name=c1 $SMALL_IMAGE /bin/sh
 assert_raises "exec_on $HOST1 c1 $CHECK_ETHWE_UP"
 
-run_on $HOST1 sudo weave --local run             -ti --name=c2 $SMALL_IMAGE /bin/sh
+run_on $HOST1 sudo COVERAGE=$COVERAGE weave --local run             -ti --name=c2 $SMALL_IMAGE /bin/sh
```
Comment (rade) [inline test/140_weave_local_test.sh]: I'd be tempted to introduce a `weave_local_on`, analogous to `weave_on`.

File: test/config.sh [hunk 2]
```diff
@@ -196,6 +205,19 @@ start_suite() {
...
     whitely assert_end
+    for host in $HOSTS; do
+        weave_stop_router $host
+    done
+}
+
+collect_coverage() {
+    host=$1
+    rm -f cover.router.prof
+    docker_on $host cp weave:/home/weave/cover.router.prof . || true
```
Comment (paulbellamy) [inline test/config.sh]: Why not copy this directly to the coverage dir/name?

File: Makefile [hunk 3]
```diff
@@ -43,7 +44,25 @@ travis: $(EXES)
...
+endif
+	@strings $@ | grep cgo_stub\\\.go >/dev/null || { \
+		rm $@; \
+		echo "\nYour go standard library was built without the 'netgo' build tag."; \
+		echo "To fix that, run"; \
+		echo "    sudo go clean -i net"; \
+		echo "    sudo go install -tags netgo std"; \
+		false; \
+	}
+
```
Comment (rade) [inline Makefile]: The duplication with the code below is annoying. @dpw can you think of quick way of reducing that?

File: Makefile [hunk 4]
```diff
@@ -35,6 +35,7 @@ WEAVE_EXPORT=weave.tar
 WEAVEEXEC_DOCKER_VERSION=1.3.1
 DOCKER_DISTRIB=prog/weaveexec/docker-$(WEAVEEXEC_DOCKER_VERSION).tgz
 DOCKER_DISTRIB_URL=https://get.docker.com/builds/Linux/x86_64/docker-$(WEAVEEXEC_DOCKER_VERSION).tgz
+COVERAGE_MODULES=./common,./common/docker,./ipam,./ipam/paxos,./ipam/ring,./ipam/space,./net,./net/address,./router
```
Comment (rade) [inline Makefile]: This list is rather prone to become outdated. Can we compute it?

File: test/config.sh (hunk 2 above)
Comment (tomwilkie) [inline test/config.sh]: docker cp takes a directory.  Its painful.

File: testing/cover/cover.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+import (
+	"fmt"
+	"os"
+	"sort"
+
+	"golang.org/x/tools/cover"
+)
+
+func merge(p1, p2 *cover.Profile) *cover.Profile {
```
Comment (paulbellamy) [inline testing/cover/cover.go]: Are overlapping blocks impossible?

File: testing/util.go [hunk 6]
```diff
@@ -50,3 +51,17 @@ func RunWithTimeout(t *testing.T, d time.Duration, f func()) {
 	}()
 	<-ch
 }
+
+// TrimTestArgs finds the first -- in os.Args and trim all args before that
+func TrimTestArgs() {
```
Comment (paulbellamy) [inline testing/util.go]: So, `flag.Args()` ?

File: Makefile (hunk 4 above)
Comment (tomwilkie) [inline Makefile]: We can computer all directories with go files in, but that would include the proxy and the nameserver.  Can't think of a way of computing all modules used by the router.

**Comment by paulbellamy** [comment]
It would be nice not to have a bunch of test gubbins littered around the `weave` script.

File: Makefile (hunk 4 above)
Comment (paulbellamy) [inline Makefile]: `go list -f '{{join .Deps "\n"}}' ./router | grep "weaveworks"`

Comment (rade) [inline Makefile]: Could we at least only list top-level dirs? @dpw no doubt can come up with a nifty one liner to turn such a list into a comma-separate list of all those dirs and their sub-dirs.

File: testing/cover/cover.go (hunk 5 above)
Comment (tomwilkie) [inline testing/cover/cover.go]: Not that I'm aware off.  Same source should produce same blocks.

**Comment by tomwilkie** [comment]
@paulbellamy re:weave script, yeah I agree, but I can't think of a better way.  Any ideas?

File: Makefile (hunk 4 above)
Comment (dpw) [inline Makefile]:
```
space:=$(subst x, ,x)
comma:=,
COVERAGE_MODULES=$(subst $(space),$(comma),$(shell find ./common ./ipam ... -type d))
```

(I'll leave implementing `find -type d` in pure make as an exercise for the reader.)

File: testing/cover/cover.go (hunk 5 above)
Comment (paulbellamy) [inline testing/cover/cover.go]: But what about sub-blocks?

File: testing/util.go (hunk 6 above)
Comment (tomwilkie) [inline testing/util.go]: Yeah I wish - for some reason it trims a little too much off.  I could investigate more if you like?

File: testing/cover/cover.go (hunk 5 above)
Comment (tomwilkie) [inline testing/cover/cover.go]: None of the go coverage code I've read would indicate that, but I'm not 100% sure TBH.

File: weave [hunk 7]
```diff
@@ -1037,6 +1042,10 @@ launch_router() {
             exit 1
         fi
     fi
+    COVERAGE_ARGS=""
+    if [ -n "$COVERAGE" ] ; then
+        COVERAGE_ARGS="-test.coverprofile=/home/weave/cover.router.prof --"
+    fi
```
Comment (rade) [inline weave]: I'd put this code in the top-level, and then in `ask_version` reset `COVERAGE_ARGS=` if the container is not the router. Removes the duplication.

**Pull request merged**